autogpt_platform/backend/backend/blocks/google/gmail.py

```diff
@@ -0,0 +1,440 @@
+
+        message = MIMEText(body)
+        message["to"] = to
+        message["subject"] = subject
+        raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode("utf-8")
+        return {"raw": raw_message}
+
+
+class GmailListLabelsBlock(Block):
+    class Input(BlockSchema):
+        credentials: GoogleCredentialsInput = GoogleCredentialsField(
+            ["https://www.googleapis.com/auth/gmail.labels"]
+        )
+
+    class Output(BlockSchema):
+        result: List[dict] = SchemaField(
+            description="List of labels",
+        )
+        error: Optional[str] = SchemaField(
+            description="Error message if any",
+        )
+
+    def __init__(self):
+        super().__init__(
+            id="3e1c2c1c-c689-4520-b956-1f3bf4e02bb7",
+            description="This block lists all labels in Gmail.",
+            categories={BlockCategory.COMMUNICATION},
+            input_schema=GmailListLabelsBlock.Input,
+            output_schema=GmailListLabelsBlock.Output,
+            test_input={
+                "credentials": TEST_CREDENTIALS_INPUT,
+            },
+            test_credentials=TEST_CREDENTIALS,
+            test_output=[
+                (
+                    "result",
+                    [
+                        {"id": "Label_1", "name": "Important"},
+                        {"id": "Label_2", "name": "Work"},
+                    ],
+                ),
+            ],
+            test_mock={
+                "_list_labels": lambda *args, **kwargs: [
+                    {"id": "Label_1", "name": "Important"},
+                    {"id": "Label_2", "name": "Work"},
+                ],
+            },
+        )
+
+    def run(
+        self, input_data: Input, *, credentials: GoogleCredentials, **kwargs
+    ) -> BlockOutput:
+        try:
+            service = GmailReadBlock._build_service(credentials, **kwargs)
+            labels = self._list_labels(service)
+            yield "result", labels
+        except Exception as e:
+            yield "error", str(e)
+
+    def _list_labels(self, service) -> List[dict]:
+        results = service.users().labels().list(userId="me").execute()
+        labels = results.get("labels", [])
+        return [{"id": label["id"], "name": label["name"]} for label in labels]
+
+
+class GmailAddLabelBlock(Block):
+    class Input(BlockSchema):
+        credentials: GoogleCredentialsInput = GoogleCredentialsField(
+            ["https://www.googleapis.com/auth/gmail.modify"]
+        )
+        message_id: str = SchemaField(
+            description="Message ID to add label to",
+        )
+        label_name: str = SchemaField(
+            description="Label name to add",
+        )
+
+    class Output(BlockSchema):
+        result: dict = SchemaField(
+            description="Label addition result",
+        )
+        error: Optional[str] = SchemaField(
+            description="Error message if any",
+        )
```

Comment:
can this block output a `None` value from the `error` output?

Reply:
isn't that an anti pattern?

Reply:
That's why I'm asking. The type annotation indicates what type can come out of an output. The current annotation is `Optional[str]` implying it can output `None`.

Reply:
O i see